File: finnssudokusite/sudoku/static/sudoku/js/solver/rules/rule_white_kropki.js
// === rule_white_kropki.js ===

import { Rule } from './rule.js';
import { EMPTY } from '../defs.js';
import { Candidates } from '../candidates.js';
import { Position } from '../position.js';

export class WhiteKropki extends Rule {
    /**
     * @param {Position} pos1
     * @param {Position} pos2
     */
    constructor(pos1, pos2) {
        super();
        this.pos1 = pos1;
        this.pos2 = pos2;
    }

    numberChanged(board, changedCell) {
        const a = board.getCell(this.pos1);
        const b = board.getCell(this.pos2);
        let changed = false;

        if (a.value !== EMPTY && b.value === EMPTY) {
            const allowed = new Candidates();
            if (a.value > 1)
                allowed.orEq(Candidates.fromNumber(a.value - 1));
            if (a.value < 9)
                allowed.orEq(Candidates.fromNumber(a.value + 1));
            const before = b.candidates.clone();
            b.candidates.andEq(allowed);
            if (!b.candidates.equals(before)) changed = true;
        }

        if (b.value !== EMPTY && a.value === EMPTY) {
            const allowed = new Candidates();
            if (b.value > 1)
                allowed.orEq(Candidates.fromNumber(b.value - 1));
            if (b.value < 9)
                allowed.orEq(Candidates.fromNumber(b.value + 1));
            const before = a.candidates.clone();
            a.candidates.andEq(allowed);
            if (!a.candidates.equals(before)) changed = true;
        }

        return changed;
    }

    candidatesChanged(board) {
        const a = board.getCell(this.pos1);
        const b = board.getCell(this.pos2);

        if (a.value !== EMPTY || b.value !== EMPTY)
            return false;

        let changed = false;

        for (const n of [...a.candidates]) {
            const hasLower = n > 1 && b.candidates.test(n - 1);
            const hasHigher = n < 9 && b.candidates.test(n + 1);
            if (!hasLower && !hasHigher) {
                a.candidates.andEq(new Candidates(~(1 << n) & Candidates.MASK_ALL));
                changed = true;
            }
        }

        for (const n of [...b.candidates]) {
            const hasLower = n > 1 && a.candidates.test(n - 1);
            const hasHigher = n < 9 && a.candidates.test(n + 1);
            if (!hasLower && !hasHigher) {
                b.candidates.andEq(new Candidates(~(1 << n) & Candidates.MASK_ALL));
                changed = true;
            }
        }

        return changed;
    }

    checkPlausibility(board) {
        const a = board.getCell(this.pos1);
        const b = board.getCell(this.pos2);
        if (a.value !== EMPTY && b.value !== EMPTY)
            return Math.abs(a.value - b.value) === 1;
        return true;
    }
}

File: src/rules/rule_white_kropki.cpp
#include "rule_white_kropki.h"
#include "../board.h"
#include "../cell.h"

namespace sudoku {

    WhiteKropki::WhiteKropki(const Position& pos1, const Position& pos2)
            : pos1_(pos1), pos2_(pos2) {}

    bool WhiteKropki::number_changed(Board& board, const Cell& changed_cell) const {
        const Cell& a = board.get_cell(pos1_);
        const Cell& b = board.get_cell(pos2_);
        bool any_changed = false;

        if (a.value != EMPTY && b.value == EMPTY) {
            Candidates allowed;
            if (a.value > 1) allowed |= Candidates(Number(a.value - 1));
            if (a.value < 9) allowed |= Candidates(Number(a.value + 1));
            Candidates& cands = board.get_cell(pos2_).candidates;
            Candidates before = cands;
            cands &= allowed;
            any_changed = (cands != before);
        }

        if (b.value != EMPTY && a.value == EMPTY) {
            Candidates allowed;
            if (b.value > 1) allowed |= Candidates(Number(b.value - 1));
            if (b.value < 9) allowed |= Candidates(Number(b.value + 1));
            Candidates& cands = board.get_cell(pos1_).candidates;
            Candidates before = cands;
            cands &= allowed;
            any_changed |= (cands != before);
        }

        return any_changed;
    }

    bool WhiteKropki::candidates_changed(Board& board) const {
        Cell& a = board.get_cell(pos1_);
        Cell& b = board.get_cell(pos2_);

        // both must be empty
        if (a.value != EMPTY || b.value != EMPTY)
            return false;

        bool any_changed = false;

        // go through all values in a, check if b has a value which is smaller or larger
        for (Number n : a.candidates) {
            bool has_lower  = n > 1 && b.candidates.test(n - 1);
            bool has_higher = n < 9 && b.candidates.test(n + 1);
            if (!has_lower && !has_higher) {
                a.candidates &= ~Candidates(n);
                any_changed = true;
            }
        }

        // go through all values in b, check if a has a value which is smaller or larger
        for (Number n : b.candidates) {
            bool has_lower  = n > 1 && a.candidates.test(n - 1);
            bool has_higher = n < 9 && a.candidates.test(n + 1);
            if (!has_lower && !has_higher) {
                b.candidates &= ~Candidates(n);
                any_changed = true;
            }
        }

        return any_changed;
    }

    bool WhiteKropki::check_plausibility(const Board& board) const {
        const Cell& a = board.get_cell(pos1_);
        const Cell& b = board.get_cell(pos2_);

        if (a.value != EMPTY && b.value != EMPTY)
            return std::abs(a.value - b.value) == 1;

        return true;
    }

} // namespace sudoku
